src/main/java/io/vertx/ext/mongo/impl/PublisherAdapter.java

```diff
@@ -150,16 +152,20 @@ private void handleOut(T item) {
     }
     handler.handle(item);
     State s;
+    int remaining;
+    int size;
     synchronized (this) {
       if (inFlight != 0) {
         return;
       }
       s = state;
+      remaining = remainingItemsInPublisherBatch;
+      size = internalQueue.size();
     }
     if (s == State.EXHAUSTED) {
       stop();
       handleEnd();
-    } else {
+    } else if (remaining == 0 && size == 0) {
```

Comment:
I think it's better not to wait for the queue to be empty before requesting more items from the publisher. Then Mongo cant start fetching more while we're processing the remaining items.